docs/development/abi.md

```diff
@@ -0,0 +1,219 @@
+# Pyodide Platform ABI
+
+## ABIs
+
+### General
+
+Shared libraries must be linked with `-sSIDE_MODULE=1` or
+
+```
+-sSIDE_MODULE=2 -sEXPORTED_FUNCTIONS=<export list>
+```
+
+Note that the name of each symbol in the list needs to be prefixed with an
+underscode. For the smallest result, it is recommended to link with:
+
+```
+-sSIDE_MODULE=2 -sEXPORTED_FUNCTIONS=["_PyInit_MyCModule1", "_PyInit_MyCModule2]
+```
+
+To force all symbols to be exported, link with `-sSIDE_MODULE=1`.
+
+If `-pthread` is used at compile or link time, the resulting libraries will not
+load.
+
+To compile Rust packages, the following flags must be passed to `rustc`:
+
+```
+-C link-arg=-sSIDE_MODULE=2 -C link-arg=-sWASM_BIGINT
+```
+
+When not invoking `rustc` directly, for instance when using cargo, set these
+into the `RUSTFLAGS` environment variable.
+
+Compiling with `-sSIDE_MODULE=1` will not work with Rust because Rust libraries
+contain a `lib.rmeta` file which is not an object file. Rust produces the
+correct list of exported symbols automatically so this should not be a problem
+in practice.
+
+On Rust libc versions older than 0.2.162, it may be necessary to pass
+`-Z link-native-libraries=no` as a `RUSTFLAG`.
+
+Pyodide is statically linked with the following libraries:
+
+- `libegl.js`
+- `libeventloop.js`
+- `libGL`
+- `libhtml5_webgl.js`
+- `libhtml5.js`
+- `liblz4`
+- `libsdl.js`
+- `libwebgl.js`
+- `libwebsocket.js`
+- `bzip2`
+- `zlib`
+
+All of these come from Emscripten ports and the versions of these libraries are
+determined by the version of Emscripten we build with. Any symbols from these
+static libraries may be used by shared libraries.
+
+### pyodide_2024_0
+
+The Emscripten version is 3.1.58. The Python version is 3.12. Python 3.12 must
+be used at build time. All shared libraries must be linked with `-sWASM_BIGINT`.
```

Comment:
Would it be worth flagging that the 3.12 pin is a result of being a "pre-Tier 3" specification?

Reply:
I don't think it has anything to do with that?

Reply:
I was mostly thinking of the weird "Tier 3 then not" status of Emscripten in Python 3.12/13/14. I guess adopting the "Python3.12 builds use the 2024_0 ABI by default" language used for the 2025_0 tag would be just as accurate.

Reply:
Since this is about packaging, I'm not sure that the CPython support tier is relevant? pypy uploads wheels to PyPI but cpython has no support policy for pypy at all.